extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java

```diff
@@ -860,8 +835,456 @@ private Table readTableInternal(
+        IcebergParquetWriteInstructions writeInstructions = verifyWriteInstructions(instructions);
+        final boolean verifySchema = writeInstructions.verifySchema().orElse(!overwrite);
+
+        final TableDefinition useDefinition;
+        if (writeInstructions.tableDefinition().isPresent()) {
+            useDefinition = writeInstructions.tableDefinition().get();
+        } else {
+            // Verify that all tables have the same definition
+            final TableDefinition firstDefinition = dhTables[0].getDefinition();
+            for (int idx = 1; idx < dhTables.length; idx++) {
+                if (!firstDefinition.equals(dhTables[idx].getDefinition())) {
+                    throw new IllegalArgumentException(
+                            "All Deephaven tables must have the same definition, else table definition should be " +
+                                    "provided when writing multiple tables with different definitions");
+                }
+            }
+            useDefinition = firstDefinition;
+            writeInstructions = writeInstructions.withTableDefinition(useDefinition);
+        }
+
+        // Try loading the table from the catalog, or create if required
+        final org.apache.iceberg.Table icebergTable;
+        final SpecAndSchema newSpecAndSchema;
+        final boolean newNamespaceCreated;
+        final boolean newTableCreated;
+        if (catalog.tableExists(tableIdentifier)) {
+            icebergTable = catalog.loadTable(tableIdentifier);
+            newSpecAndSchema = createSpecAndSchema(useDefinition, writeInstructions);
+            newNamespaceCreated = false;
+            newTableCreated = false;
+            if (verifySchema) {
+                try {
+                    if (overwrite) {
+                        verifyOverwriteCompatibility(icebergTable.schema(), newSpecAndSchema.schema());
+                        verifyOverwriteCompatibility(icebergTable.spec(), newSpecAndSchema.partitionSpec());
+                    } else {
+                        verifyAppendCompatibility(icebergTable.schema(), useDefinition, writeInstructions);
+                        verifyAppendCompatibility(icebergTable.spec(), useDefinition, writeInstructions);
+                    }
+                } catch (final IllegalArgumentException e) {
+                    throw new IllegalArgumentException("Schema verification failed. Please provide a compatible " +
+                            "schema or disable verification in the Iceberg instructions. See the linked exception " +
+                            "for more details.", e);
+                }
+            }
+        } else if (writeInstructions.createTableIfNotExist()) {
+            newNamespaceCreated = createNamespaceIfNotExists(tableIdentifier.namespace());
+            newSpecAndSchema = createSpecAndSchema(useDefinition, writeInstructions);
+            icebergTable = createNewIcebergTable(tableIdentifier, newSpecAndSchema, writeInstructions);
```

Comment:
Can this stuff be done as a transaction? Possibly with removing/adding the data files, as well?

Reply:
You said that this is hard for software engineering reasons. I suspect we can refactor to allow for it.